File: Data Structures & Algorithms/reverse-a-linked-list/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if (head == NULL) {
            return NULL;
        }
        return flipper(NULL, head);
    }

    ListNode* flipper(ListNode *prev, ListNode *curr) {
        if (curr->next != NULL) {
            ListNode* oldNext = curr->next;
            curr->next = prev;
            return flipper(curr, oldNext);
        }
        // base case
        curr->next = prev;
        return curr;
    }
};

File: Data Structures & Algorithms/reorder-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        stack = []
        curr = head
        size = 0
        while curr:
            stack.append(curr)
            curr = curr.next
            size += 1

        curr = head

        for _ in range(size // 2):
            friend = stack.pop()
            friend.next = curr.next
            curr.next = friend
            curr = friend.next
        curr.next = None